float8_experimental/float8_linear_utils.py

```diff
@@ -322,3 +332,47 @@ def inner_func():
     for child in fp8_layers:
         # Set a flag to signal amaxes/scales are ready
         child.amax_and_scale_synced = True
+
+
+def precompute_float8_weights(module: nn.Module) -> None:
+    from torch.distributed._tensor import DTensor
+
+    if any(isinstance(m, Float8Linear) for m in module.modules()):
+        raise NotImplementedError(
+            f"Only supports Float8DynamicLinear, not Float8Linear"
+        )
+    float8_linears: List[Float8DynamicLinear] = [
+        m
+        for m in module.modules()
+        if isinstance(m, Float8DynamicLinear)
+        and isinstance(m.weight, DTensor)
+        and isinstance(m.weight._local_tensor, Float8DynamicLinearWeightTensor)
```

Comment:
do I understand correctly how this currently works?

1. single GPU: Float8DynamicLinear is initialized with weight of type nn.Parameter
2. when distributed code is initialized, Float8DynamicLinear instances get their weight wrapped with DTensor
3. the wrapped DTensor weights have a _local_tensor attribute of type Float8DynamicLinearWeightTensor

? 

if I understood that correctly, one confusion I have is that it seems that we are defining `Float8DynamicLinear` and `Float8DynamicLinearWeightTensor` together, but they need a separate piece of code (which lives outside of this repo) to know how to work together.  Would be great to see if we can define the interaction more clearly here.

Reply:
is there anything stopping us from just wrapping the weight with `Float8DynamicLinearWeightTensor Float8DynamicLinearWeightTensor` regardless of whether distributed is used or not?

Reply:
> is there anything stopping us from just wrapping the weight with Float8DynamicLinearWeightTensor Float8DynamicLinearWeightTensor regardless of whether distributed is used or not?

I think in an ideal world where tensor subclasses had parity with `torch.Tensor`, then we should unify the code paths and always use the subclass.

Some things with the subclass:
- `torch.optim` optimizers and `clip_grad_norm_` no longer default to `foreach=True` (which lead to silent performance degradation)
- Ops now go through `__torch_dispatch__`, which currently runs up to two `tree_map_only` calls, incurring extra CPU overheads

For FSDP2, we are accepting these concerns since we do not have much choice. For single-GPU, if users do not care about the FSDP all-gather, then they may not want to opt into these concerns.

> do I understand correctly how this currently works?

Your understanding looks right to me! 

> if I understood that correctly, one confusion I have is that it seems that we are defining Float8DynamicLinear and Float8DynamicLinearWeightTensor together, but they need a separate piece of code (which lives outside of this repo) to know how to work together. Would be great to see if we can define the interaction more clearly here.

Is that separate piece of code referring to `DTensor` specifically? Would you prefer a comment to document this in the file or something else?

Reply:
Also the DTensor in question is not a result of TP/SP but rather the ones used in the implementation of FSDP2 right? Do you think other subclasses looking to interop correctly with fsdp2 will also need similiar logic?

Reply:
> Also the DTensor in question is not a result of TP/SP but rather the ones used in the implementation of FSDP2 right?

Yep!

> Do you think other subclasses looking to interop correctly with fsdp2 will also need similiar logic?

I think the particular issue is related to wrapper subclasses (including `DTensor`). Wrapper subclasses do not respect the `isinstance()` checks that would otherwise work on their inner tensor(s).

I am not sure what a good solution here would be since some wrapper subclasses like `DTensor`always have 1 inner tensor, in which case forwarding a type/isinstance check to the inner tensor could make sense, while other wrapper subclasses have several inner tensors and forwarding such checks to inner tensors may not make sense.

Reply:
I think the rest of the code looks good, curious if @vkuzo has more thoughts on this aspect